openprescribing/data/rxdb/labelled_matrix.py

```diff
@@ -24,3 +26,82 @@ class LabelledMatrix:
 
     def __post_init__(self):
         assert self.values.shape == (len(self.row_labels), len(self.col_labels))
+
+    def group_rows_by_label(self, row_label_map: tuple[LabelPair, ...]):
+        """
+        Produce a new LabelledMatrix by mapping the existing set of rows to a new set
+        using the supplied mapping, grouping together (i.e. summing column-wise) any
+        rows which are mapped to the same output label.
+
+        The mapping should be supplied as a sequence of pairs:
+
+            old_label_1 -> new_label_1
+            old_label_2 -> new_label_1
+            old_label_3 -> new_label_2
+            old_label_4 -> new_label_2
+            ...
+
+        Note that it's possible to just select a subset of rows without doing any
+        grouping simply by ensuring that each input row maps to at most one output row:
+
+            old_label_1 -> new_label_1
+            old_label_4- > new_label_2
+            old_label_9- > new_label_3
+            ...
+
+        """
+        row_grouper, new_row_labels = create_row_grouper(self.row_labels, row_label_map)
+        grouped_values = row_grouper(self.values)
+        return self.__class__(grouped_values, new_row_labels, self.col_labels)
+
+
+def create_row_grouper(input_labels, label_map):
+    """
+    Construct a function which can efficiently sum together arbitrary groups of rows in a matrix
+
+    All the number crunching here is done using a SciPy sparse matrix dot product, hence
+    the efficiency. We just need to construct the appropriate sparse matrix. As with the
+    `get_grouped_sum_ndarray` function, this is such a core operation to OpenPrescribing
+    that it makes sense to optimise it even at the cost of some linear algebra.
+    """
+    input_label_index = {label: i for i, label in enumerate(input_labels)}
+
+    # Build a mapping from output labels to the list of input row indexes which are
+    # mapped to that label
+    output_labels = {}
```

Comment:
Out of interest, why not use a defaultdict?

Reply:
I think because we need to ensure the key exists even if the corresponding list is empty (as part of the graceful handling of missing labels). And in an earlier iteration I ended up with the line:
```py
output_labels[output_label]
```
which was correct but just looked weird, so I switched to using `setdefault()`

Given that I'm now assigning the result of that to a variable the defaultdict version would be fine. Happy to switch to that if you think it would be clearer (even by just a tiny margin).